### PR TITLE
SOAP backchannel logout for SAML protocol

### LINKED ISSUE #16293: SAML SOAP Backchannel Single-logout
### Description

Keycloak currently supports SAML logout through POST and Redirect bindings (front channel). If backchannel logout is selected, then the binding is kept, but the request is directly sent (POST/GET) to the SAML client logout URL, and some redirections are followed if encoutered.
The aim of this enhancement is to add a real backchannel (with SOAP binding) logout capability to SAML client registered in Keycloak.
Additionally, it also adds the capability for Keycloak to receive logout requests sent by any SAML client over SOAP binding.

### Discussion

https://github.com/keycloak/keycloak/discussions/9548

### Motivation

Currently, the SAML backchannel logout is not compliant with the SAML protocol. This enhancement adds the SOAP backchannel logout as described in the SAML protocol, but keeps the "legacy" backchannel logout with POST and Redirect binding for backward compatibility purpose. 

### Details

A new SAML client attribute is introduced (Logout Service SOAP Binding URL): if this URL is set on a client, and the client is configured with backchannel logout, then the SOAP binding is used for sending the SAML logout request. Otherwise, the behavior stays the same as currenty implemented.

### PR DESCRIPTION
**Pull request by lagess**
Closes #16293

- add support to send a SAML Logout request over SOAP binding to any compliant client
- add support for a SAML client to trigger a Single logout process over SOAP binding

### CODE REVIEW

**Comment by jonkoops** [comment]
@mhajas have you had some time to take a look at this? We have a related PR on the Admin Console we'd like to get merged.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:301]: If I understand correctly, this test is testing this from spec:
```
Synchronous Bindings (Back-Channel)
Alternatively, the session participant MAY use a synchronous binding, such as the SOAP binding
[SAMLBind], to send the request directly to the identity provider. The identity provider SHOULD then
propagate any required logout messages to additional session participants as required using a
synchronous binding. The requester MUST authenticate itself to the identity provider, either by signing
the <LogoutRequest> or using any other binding-supported mechanism.
```
the last part `The requester MUST authenticate itself ....`.

This test is currently passing because of [this check](https://github.com/keycloak/keycloak/blob/main/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L301-L303). 

I am wondering, should this test pass also for `SAML_CLIENT_ID_SALES_POST` that does not have `"saml.client.signature"` set to `true`? It does not at the moment, should we add some check that requires a signature for `SOAP` binding regardless of the client settings?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (mhajas) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:269]: Can you elaborate on why this is needed?

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you for the PR @lagess! I went through the changes and added a few comments, mostly nitpicks, but I also added one question regarding the signature requirement for SOAP logout.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:619]: I do not think we should: the parameters in this function do no make much sense and are mainly related to the legacy behavior of the relationship between binding and front/backchannel.
If SOAP Binding is requested, it must be backchannel.
We could check for frontchannel + soap-binding and throw some exception in that case. But if we hit the exception, then it means that there is an implementation issue.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:621]: I think it makes sense, indeed.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (lagess) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:301]: I'd say that the authentication of the requester is mandatory, but the means is not defined. It can be that logout requests are signed, but it could also be that there are some mutual TLS established on the link between the client and the IDP. Thus I would not make the message-level signature mandatory, but rather let the integrator decide how authentication is achieved through configuration of the client (`saml.client.signature`).

What do you think?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (lagess) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:269]: This is to avoid that the cookie is used for retrieving the user session when the logout request is received. I want to ensure that the session index contained in the logout request is used for determining the session to disconnect (i.e. that [this code](https://github.com/keycloak/keycloak/blob/022d2864a6fa71bd97c6f4ac86de178f8bdcfbd6/services/src/main/java/org/keycloak/protocol/saml/SamlService.java#L585) is used).

**Comment by lagess** [comment]
Thanks, @mhajas, I've addressed your comments.
Let me know if something is still not clear!

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:298]: ok, makes sense to me. Thanks for the explanation.

Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:298]: I agree that authentication is out of the SAML SOAP Binding scope. I'm also comfortable with the current change in this aspect. Signatures and https should be enabled if a decent level of security is wanted (it's the same for all the SAML configuration).

But we have to keep in mind that clients can have weird things to allow the call from the server to execute the logout in then (for example basic auth, certificates,...). With the current implementation the server just sends the SOAP message with/without signature (nothing more) and this fact should be clear (maybe doc or something). I don't like either that SOAP is using default [SOAPConnectionFactory](https://github.com/keycloak/keycloak/blob/20.0.3/services/src/main/java/org/keycloak/protocol/saml/profile/util/Soap.java#L207-L209) to perform the SOAP call. It means using default `HttpURLConnection` instead of `SimpleHttp` (proxies, certs,... are not taken into account). But this is another issue, @mhajas WDYT? Do we file another issue for that?

The same for the server, it accepts logout requests with or without signature. Although prod configurations needs https and signatures for SAML to be minimally secure.

**Review by rmartinc — COMMENTED** [review]
@lagess At least checking the signature is needed in the synch response if the client requires signatures.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:769]: We can indeed validate the response. But then what would you expect in terms of impact? As the SLO feature is mainly a best-effort move, then validating the response will not impact what happens next during the SLO. To me, the main consequence of receiving a bad response from the client will end up being an additional warning entry in the logs.

Comment (rmartinc) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:769]: Yep, you are right. The logout is already done and any exception will produce just a warning of some kind. But don't you think is important knowing that the endpoint is not signing the response? Even to advise the configurator there is something weird here?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:298]: @rmartinc this change sounds correct, however, I am not sure what changes we would require to make this happen. For sure, we can create an issue and then prioritize.

File: services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java
Comment (lagess) [reply services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java:769]: Yes, it can indeed be useful to know that something is not appropriately configured, or behaving badly.
I'll check how to improve on that.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java
Comment (rmartinc) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:298]: @mhajas I created https://github.com/keycloak/keycloak/issues/17139 and assigned to me. I'll work on it when I have some time.

Comment (mhajas) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/saml/LogoutTest.java:298]: Thank you @rmartinc

**Review by mhajas — CHANGES_REQUESTED** [review]
@rmartinc Could you please approve if you are happy with this? 
@lagess Could you please rebase? We removed the old admin console from the repository so the changes in `themes/*` should be removed.

**Comment by rmartinc** [comment]
@mhajas I think that @lagess agreed to perform the signature check in a comment, so I'm waiting for those changes.

**Comment by lagess** [comment]
@rmartinc Yes I'm currently adding a log entry when the signature check fails, should be ready soon. BTW, I cannot really test this, as checking for a log entry is a bit tedious to do...

**Comment by lagess** [comment]
@rmartinc I've rebased the PR and added the log entry when the response signature is invalid.
For now, I've kept the fixup for easier reviewing. Once approved, I'll squash into one commit.

**Review by rmartinc — CHANGES_REQUESTED** [review]
@lagess The verification LGTM but two more things:

* I detected that there are SAML tests that fail. My feeling is that the problem is the getLogoutServiceUrl is returning the AdminURL/ManagementURL for a soap binding ([here](https://github.com/keycloak/keycloak/blob/28591775fd36735fd9628fb2d101d8cda74509d4/services/src/main/java/org/keycloak/protocol/saml/SamlProtocol.java#L635-L636)). And the logout is trying a soap logout when the legacy post should be done. Maybe that if should be modified to not return the management url in case of a soap binding.

```
[ERROR]   SAMLServletAdapterTest.testDifferentCookieName:1540->AbstractSAMLServletAdapterTest.checkLoggedOut:99 » ConditionTimeout Assertion condition defined as a lambda expression in org.keycloak.testsuite.util.URLAssert that uses java.lang.String URL expected to begin with: https://localhost:8543/auth/realms/demo/login-action ; actual URL: http://localhost:8280/different-cookie-name/ within 10 seconds.
```

* There is a new option for SAML `saml_single_logout_service_url_soap` that should be added to the UI. Take a look to [FineGrainSamlEndpointConfig](https://github.com/keycloak/keycloak-ui/blob/20.0.4/apps/admin-ui/src/clients/advanced/FineGrainSamlEndpointConfig.tsx#L59-L75) and create a group for the new property (and add the English texts). Send the tiny PR to the UI project too.

Thanks!

**Comment by lagess** [comment]
@rmartinc Thanks for the heads up about the failing tests. I've modified the getLogoutServiceUrl according to your suggestion, and it helps with the adapter tests.
About the admin UI: I had already opened 2 additional PRs for the admin console and the documentation (https://github.com/keycloak/keycloak-documentation/pull/1740, https://github.com/keycloak/keycloak-ui/pull/4138). Is it what you are requesting?

**Review by rmartinc — CHANGES_REQUESTED** [review]
> About the admin UI: I had already opened 2 additional PRs for the admin console and the documentation (https://github.com/keycloak/keycloak-documentation/pull/1740, https://github.com/keycloak/keycloak-ui/pull/4138). Is it what you are requesting?

Yes, exactly, I had not checked you had already sent a PR for the UI. :smile: 

For me it's almost OK, just the minor comment to not return an empty/blank logout URL as sometimes the console does. With that the PR LGTM.

Thanks!

**Comment by lagess** [comment]
@rmartinc I've applied the fix as suggested. There is some `ObjectUtil.isBlank(...)` that we could have used, but I've kept the way you suggested because it is done this way for the other bindings.

**Review by rmartinc — APPROVED** [review]
@mhajas LGTM now!

**Review by mhajas — APPROVED** [review]
Thanks for the contribution @lagess

**Pull request merged**
